File: setupFirebase.py
from installPackages import install_packages, toast
import os
from Generics import ReplaceFileContent
from Constants import CodeConstants, FileConstants
from toast import colors
import shutil

def init_firebase(directory:str, project_name: str):
    # Install Firebase App
    install_packages(['@react-native-firebase/app'])
    
    # Replace the contents of the android/build.gradle file
    ReplaceFileContent(f'{directory}/android/build.gradle', 
                       [CodeConstants.FIREBASE.BUILD_GRADLE_OLD], [CodeConstants.FIREBASE.BUILD_GRADLE_NEW])

    # Replace the contents of the android/app/build.gradle file
    ReplaceFileContent  (f'{directory}/android/app/build.gradle', 
                         [CodeConstants.FIREBASE.ANDROID_BUILD_GRADLE_OLD], [CodeConstants.FIREBASE.ANDROID_BUILD_GRADLE_NEW])

    # Replace the contents of the AppDelegate.mm file
    ReplaceFileContent(f'{directory}/ios/{project_name}/AppDelegate.mm', 
                       [CodeConstants.FIREBASE.APPDELEGATE_OLD_IMPORT, CodeConstants.FIREBASE.APPDELEGATE_OLD_CONFIGURE], 
                       [CodeConstants.FIREBASE.APPDELEGATE_NEW_IMPORT, CodeConstants.FIREBASE.APPDELEGATE_NEW_CONFIGURE])

    # Replace the contents of the Podfile file
    ReplaceFileContent(f'{directory}/ios/Podfile', 
                       [CodeConstants.FIREBASE.PODFILE_OLD], [CodeConstants.FIREBASE.PODFILE_NEW])
    
    toast('Firebase initialized successfully, Dont forget to add GoogleService-Info.plist and google-services.json', type='info')


def init_messaging(directory:str, project_name: str):

    # Install Firebase Messaging
    install_packages(['@react-native-firebase/messaging', '@notifee/react-native'])

    source_file_base_path = '/Users/aashirraza/Desktop/Tests/react-native-generator-functions/'

    print(colors.YELLOW + 'Please follow the instructions at' + colors.YELLOW_BOLD + ' https://rnfirebase.io/messaging/usage/ios-setup#enable-push-notifications' + colors.END)

    # make a copy of the file
    shutil.copy(f'{source_file_base_path}/notificationServiceTemplate.ts', '/Users/aashirraza/Desktop/Tests/react-native-generator-functions/notificationService.ts')

    # Replace the contents of the notificationService.ts file
    ReplaceFileContent(f'{source_file_base_path}/notificationService.ts', ['genericName_notification_channel_id'], [f'{project_name}_notification_channel_id'])

    # Copy the file to the project
    shutil.copy(f'{source_file_base_path}/notificationService.ts', f'{directory}/src/services')

    # Remove the original file
    os.remove(f'{source_file_base_path}/notificationService.ts')

    # Replace the contents of the index.js file
    ReplaceFileContent(f'{directory}/index.js', [CodeConstants.FIREBASE.INDEX_JS_OLD], [CodeConstants.FIREBASE.INDEX_JS_NEW])
    
def main():
    # Get the current working directory and project name
    current_directory = os.getcwd()
    project_name = current_directory.split('/')[-1]

    # Check if the current directory contains the necessary files or folders
    if not all(os.path.exists(os.path.join(current_directory, f)) for f in FileConstants.REQUIRED_FOLDERS):
        print("Error: Please execute this script from the root folder of your project directory.")
        return
    
    while True:
        print("Press 1 for basic init")
        print("Press 2 for messaging init")
        print("Press 3 for combine init")
        print("Press 4 to exit")

        choice = input("Enter your choice: ")

        if choice == '1':
            init_firebase(directory=current_directory, project_name=project_name)
        elif choice == '2':
            init_messaging(directory=current_directory, project_name=project_name)
        elif choice == '3':
            init_firebase(directory=current_directory, project_name=project_name)
            init_messaging(directory=current_directory, project_name=project_name)
        elif choice == '4':
            print("Exiting...")
            break
        else:
            print("Invalid choice. Please enter a valid option.")    

if __name__ == "__main__":
    main()

File: src/SetupFirebase.js
import fs from "fs";
import { copyFile, installPackages, replaceFileContent } from "./Utils.js";
import { Toast } from "./Toast.js";
import {
  FileConstants,
  FirebaseConstants,
  ToastMessages,
  ToastTypes,
} from "./Constants.js";
import ora from "ora";
import pkg from "enquirer";
const { Input } = pkg;

async function createNotificationService(directory, both = false) {
  const spinner = ora({
    text: "Creating notification service...",
    spinner: "aesthetic",
    prefixText: both ? "[8/9]" : "[3/4]",
  }).start();

  try {
    // Copy the file to the project asynchronously
    await copyFile(
      FileConstants.NOTIFICATION_TEMPLATE_COPY,
      `${directory}/src/services/notificationService.ts`
    );
    spinner.succeed("Notification service created");
  } catch (error) {
    console.log(error);
    spinner.fail("Failed to create notification service");
    throw error;
  }
}

const initFirebase = async (directory, projectName, both = false) => {
  const firebasePackages = ["@react-native-firebase/app"];
  try {
    await installPackages(
      firebasePackages,
      false,
      "Firebase App",
      both ? "[1/9]" : "[1/5]"
    );

    const androidBuildGradlePath =
      FileConstants.PATH_TO_GRADLE.format(directory);
    const androidAppBuildGradlePath =
      FileConstants.PATH_TO_ANDROID_GRADLE.format(directory);
    const appDelegatePath = FileConstants.PATH_TO_APP_DELEGATE.format(
      directory,
      projectName
    );
    const podfilePath = FileConstants.PATH_TO_PODFILE.format(directory);

    await replaceFileContent(
      androidBuildGradlePath,
      [FirebaseConstants.BUILD_GRADLE_OLD],
      [FirebaseConstants.BUILD_GRADLE_NEW],
      both ? "[2/9]" : "[2/5]"
    );

    await replaceFileContent(
      androidAppBuildGradlePath,
      [FirebaseConstants.ANDROID_BUILD_GRADLE_OLD],
      [FirebaseConstants.ANDROID_BUILD_GRADLE_NEW],
      both ? "[3/9]" : "[3/5]"
    );

    await replaceFileContent(
      appDelegatePath,
      [
        FirebaseConstants.APPDELEGATE_OLD_IMPORT,
        FirebaseConstants.APPDELEGATE_OLD_CONFIGURE,
      ],
      [
        FirebaseConstants.APPDELEGATE_NEW_IMPORT,
        FirebaseConstants.APPDELEGATE_NEW_CONFIGURE,
      ],
      both ? "[4/9]" : "[4/5]"
    );

    await replaceFileContent(
      podfilePath,
      [FirebaseConstants.PODFILE_OLD],
      [FirebaseConstants.PODFILE_NEW],
      both ? "[5/9]" : "[5/5]"
    );

    Toast(ToastMessages.FIREBASE_INIT_SUCCESS, ToastTypes.SUCCESS);

    return true;
  } catch (error) {
    Toast(ToastMessages.FIREBASE_INIT_ERROR, ToastTypes.ERROR);
    return false;
  }
};

const initMessaging = async (directory, projectName, both = false) => {
  const messagingPackages = [
    "@react-native-firebase/messaging",
    "@notifee/react-native",
  ];
  try {
    await installPackages(
      messagingPackages,
      false,
      "Firebase Messaging",
      both ? "[6/9]" : "[1/4]"
    );

    Toast(ToastMessages.FIREBASE_MESSAGING_INIT, ToastTypes.INFO);

    // make a copy of the file
    fs.copyFileSync(
      FileConstants.NOTIFICATION_TEMPLATE_PATH,
      FileConstants.NOTIFICATION_TEMPLATE_COPY
    );

    // Replace the contents of the notificationService.ts file
    await replaceFileContent(
      FileConstants.NOTIFICATION_TEMPLATE_COPY,
      ["genericName_notification_channel_id"],
      [`${projectName}_notification_channel_id`],
      both ? "[7/9]" : "[2/4]"
    );

    await createNotificationService(directory, both);

    // Remove the original file
    fs.unlinkSync(FileConstants.NOTIFICATION_TEMPLATE_COPY);

    // Replace the contents of the index.js file
    await replaceFileContent(
      `${directory}/index.js`,
      [FirebaseConstants.INDEX_JS_OLD],
      [FirebaseConstants.INDEX_JS_NEW],
      both ? "[9/9]" : "[4/4]"
    );

    return true;
  } catch (error) {
    Toast(ToastMessages.FIREBASE_MESSAGING_ERROR, ToastTypes.ERROR);
    return false;
  }
};

const main = async () => {
  // Get the current working directory and project name
  const currentDirectory = process.cwd();
  const projectName = currentDirectory.split("/").pop();

  if (
    !FileConstants.REQUIRED_FOLDERS.every((f) =>
      fs.existsSync(`${currentDirectory}/${f}`)
    )
  ) {
    Toast(ToastMessages.REQUIRED_FOLDERS_MISSING, ToastTypes.ERROR);
    process.exit(1);
  }

  console.log("\nPress 1 for basic init");
  console.log("Press 2 for messaging init");
  console.log("Press 3 for combine init");
  console.log("Press any key to exit to exit\n");

  const prompt = new Input({
    name: "choice",
    message: "Enter your choice:",
  });
  const choice = await prompt.run();

  switch (choice) {
    case "1":
      await initFirebase(currentDirectory, projectName);
      break;
    case "2":
      await initMessaging(currentDirectory, projectName);
      break;
    case "3":
      await initFirebase(currentDirectory, projectName, true);
      await initMessaging(currentDirectory, projectName, true);
      break;
    default:
      console.log("Exiting...");
      process.exit(0);
  }
};

main();
